Check client for permission inconsistency

cravtos opened the pull request:
Насколько хорошим решением будет сделать так, как я сделал?

Кажется не особо правильным, что функции возвращают Promise.
И в `sendData()` кажется, что получилось нагромождение.

Идёт с: https://github.com/antibot-dev-team/antibot-backend/pull/9

cravtos committed:
Check client for permission inconsistency

manmolecular commented on src/properties.ts:
```diff
@@ -102,15 +103,31 @@ class Properties {
+    return navigator.permissions.query({name: 'notifications'})
+      .then(function(permissionStatus: PermissionStatus) {
+        return Notification.permission === 'denied' && permissionStatus.state === 'prompt';
+      });
```
Можно немного проще:
```js
    return navigator.permissions.query({name: 'notifications'})
      .then(permissionStatus => Notification.permission === 'denied' && permissionStatus.state === 'prompt');
```

manmolecular replied:
Второй return не нужен, т.к. мы упростили всё до стрелочного синтаксиса, а выше ты уже возвращаешь промис

manmolecular commented on src/properties.ts:
```diff
@@ -102,15 +103,31 @@ class Properties {
+          webdriver: this.getWebdriver(),
+          has_window_chrome: this.hasWindowChrome(),
+        };
+      }.bind(this));
```
Можно упростить с async/await, чтобы избежать нагромождения:
```js
  async collect(): Promise<ClientProperties> {
    const inconsistent_permissions = await this.checkPermissions();

    return {
      inconsistent_permissions: inconsistent_permissions,
      languages: this.getLanguages(),
      plugins: this.getPlugins(),
      custom_window: this.getWindowCustomProperties(),
      ua: this.getUserAgent(),
      webdriver: this.getWebdriver(),
      has_window_chrome: this.hasWindowChrome(),
    };
  }
```

manmolecular replied:
И переменная `inconsistent_permissions` должна быть `inconsistentPermissions` - мы используем underscore только для общения с JSON-структурами (т.к. на бэкэнд стороне ожидаем underscore-style), а внутри кода, во всем, что не попадает в JSON напрямую делаем camelCase, стандарт для js

Т.е.
```js
return {
  inconsistent_permissions: inconsistentPermissions
...
};
```
Первое - ключ в JSON с underscore, второе - значение нашей переменной в camelCase

manmolecular commented on src/agent.ts:
```diff
@@ -40,35 +40,39 @@ class Agent {
+        properties: collected_properties
+      };
+    });
   }
```
Можно упростить:
```js
  /**
   * Prepare body with client's data
   */
  async prepareRequestBody(): Promise<AnalyzeRequest> {
    // TODO/Brainstorm: Think about request format
    //  Questions here:
    //  1. Which parameters are required to correctly detect bots? UA is enough?
    //  2. Which window.*, document.* objects we need?

    return {
      properties: await this._propertiesCollector.collect()
    };
  }
```

Я бы хотел сказать что нам эта функция (подготовки тела) вообще по сути не нужна, т.к. она у нас сейчас просто оборачивает вызов collect, но тем не менее её хотелось бы оставить как враппер для будущих проверок, т.е. для их подготовки и т.д.

Возможно, тут будет несколько await и т.д.

manmolecular commented on src/agent.ts:
```diff
@@ -40,35 +40,39 @@ class Agent {
+        }.bind(this))
+        .then(function (response: AnalyzeResponse) {
+          this.handleResponse(response);
+        }.bind(this)));
```
Я бы тут последнюю `);` перенес на новую строку

manmolecular replied:
Саму функцию я бы оставил так, как есть у тебя. Немного объемно, но сама цепочка действий прослеживается - подготовить тело, подготовить запрос, затем отправить всё это и обработать ответ. Можно как-то разнести, но я бы сейчас не стал

manmolecular reviewed (changes requested):
Для нашего текущего решения - нормально. Т.е., если говорить более широко, то у нас потенциально могла бы быть (в будущем, при большом количестве разношерстных проверок, синхронных и асинхронных) одна async-функция (допустим, та же `prepareRequestBody` или `collect`), которая await-ила бы различные async-проверки внутри себя, складывала их в одно тело и возвращала промис, который мы затем бы резолвили - то есть, нам не нужно было бы делать множественный `.then()` для всех, или там строить огромный чейн, мы могли бы это объединить, но сейчас нам это не нужно - у нас это фактически частный случай с navigator query.

Функции возвращают Promise, по коду всё ок - под "не особо правильно" ты имеешь в виду что? С точки зрения кода, с точки зрения архитектурной, или с какой-то ещё? Можно где-то переписать на async/await синтаксис, но это самой логики исполнения сейчас не изменит, т.е. в этом же синтаксическом сахаре будет работать тот же механизм.

С `sendData()` отписал в комментарии к функции.

Можно немного упростить, и я сделаю пере-ревью.

manmolecular commented:
Ещё одно замечание - наверное, нужно проверять на consistency, а не inconsistency. Т.е. свойство должно быть не инвертировано при отправке, а отправляться как есть. Если consistent, то true, если insonsistent, то false.

cravtos committed:
Simplify functions with promises.

Also rename inconsistent_permissions to consistent_permissions (inverse
value correspondingly)

cravtos commented:
Исправил замечания / сделал упрощения в одном коммите: f8d6d1f

По большей части сомнения были по поводу нагромождения промисов. Выглядело не очень красиво. Но после упрощения (особенно после штук await) стало намного лучше.

cravtos commented on src/agent.ts:
```diff
@@ -40,35 +40,39 @@ class Agent {
-    }.bind(this)).then(function (response: AnalyzeResponse) {
-      this.handleResponse(response);
-    }.bind(this));
+    this.prepareRequestBody()
+      .then(body => {
+        return {
+          method: 'POST',
+          body: JSON.stringify(body),
+          headers: {
+            'Content-Type': 'application/json'
+          }
+        };
+      })
+      .then(options => fetch(Endpoints.analyze, options)
+        .then(function (response: Response) {
+          return response.json();
+        }.bind(this))
+        .then(function (response: AnalyzeResponse) {
+          this.handleResponse(response);
+        }.bind(this))
+      );
```
Может здесь после `fetch()` тоже использовать стрелочные функции?

Вроде
```javascript
.then(options => fetch(Endpoints.analyze, options)
        .then(response => {
          return response.json();
        })
        .then(response => {
          this.handleResponse(response);
        })
      );
```

или даже

```javascript
.then(options => fetch(Endpoints.analyze, options)
        .then(async response => {
          this.handleResponse(await response.json());
        })
      );
```

Как я понял в стрелочных функциях ещё и не нужно биндить `this`.

manmolecular replied:
Да, конечно. Но я бы второй вариант не стал здесь использовать - мне кажется, выглядит немного перегруженно на первый взгляд? Т.е. вызывает может быть легкий ступор в первые секунды.

Может быть вот так?
```js
      .then(options => fetch(Endpoints.analyze, options)
        .then(response => response.json())
        .then(responseBody => this.handleResponse(responseBody))
        .catch(error => console.log(error))
      );
```

В остальном на твой вкус, но поддерживаю что тут тоже надо упростить.

manmolecular replied:
@Cravtos

cravtos replied:
Использовал твой вариант с однострочными функциями -- b0b3965, но без ловли ошибок.

Ещё думаю о таком упрощении:
```javascript
.then(body => ({
        method: 'POST',
        body: JSON.stringify(body),
        headers: {
          'Content-Type': 'application/json'
        }
      }))
```
вместо
```javascript
.then(body => {
        return {
          method: 'POST',
          body: JSON.stringify(body),
          headers: {
            'Content-Type': 'application/json'
          }
        };
      })
```

Что скажешь?

manmolecular replied:
Поддерживаю

cravtos replied:
Готово -- 29e0786

cravtos committed:
Catch for TypeError in checkPermissions()

Tested on Firefox 45: trying to access navigator.permission raises
TypeError.

cravtos committed:
Fix typo

cravtos committed:
Use arrow functions in sendData()

cravtos committed:
Remove redundant return in sendData()

manmolecular reviewed (approved):
LGTM
